### PR TITLE
Fix UnauthorizedLogin errors during buildbot start

### PR DESCRIPTION
**Pull request by rjarry**
This should fix issue #3462.

I did not add any news fragments for now. I'd like some feedback on whether this is the correct way to go or if I have it all backwards.

### CODE REVIEW

**Comment by rjarry** [comment]
I changed the approach following @tardyp advice. It works :)

File: master/buildbot/master.py [hunk 1]
```diff
@@ -334,53 +334,29 @@ def sendBuildbotNetUsageData(self):
...
-        # a time, and alerting the user if the reconfig takes too long
-        if self.reconfig_active:
-            log.msg("reconfig already active; will reconfig again after")
-            self.reconfig_requested = True
-            return
-
-        self.reconfig_active = self.reactor.seconds()
-        metrics.MetricCountEvent.log("loaded_config", 1)
-
-        # notify every 10 seconds that the reconfig is still going on, although
```
Comment (tardyp) [inline master/buildbot/master.py]: I think you should still keep the long reconfig warning timer.

Also, if you send 10 SIGHUP in a row, you'll get 10 reconfig, while the previous implementation, only 2

File: master/buildbot/master.py [hunk 2]
```diff
@@ -334,53 +334,29 @@ def sendBuildbotNetUsageData(self):
...
-
-        d.addErrback(log.err, 'while reconfiguring')
-
-        return d  # for tests
+        try:
+            yield self.initLock.acquire()
+            metrics.MetricCountEvent.log("loaded_config", 1)
+            yield self.doReconfig()
+        finally:
+            yield self.initLock.release()
```
Comment (tardyp) [inline master/buildbot/master.py]: maybe the better implementation is to just keep reconfig as is, and to change doReconfig to take the lock

**Comment by rjarry** [comment]
This should look better now. I added a newsfragment and a simple unit test for pbmanager.

**Comment by rjarry** [comment]
BTW, the smoke tests failed but it looks unrelated to these changes:

```
...
error An unexpected error occurred: "https://registry.yarnpkg.com/async/-/async-1.5.2.tgz: ENOENT: no such file or directory, utime '/home/buildbot/.cache/yarn/v1/npm-async-1.5.2-ec6a61ae56480c0c3cb241c95618e20892f9672a/README.md'".
info If you think this is a bug, please open a bug report with the information provided in "/buildbot/buildbot-job/build/www/waterfall_view/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Running command: /buildbot/buildbot-job/build/www/waterfall_view/node_modules/.bin/gulp prod --notests
error An unexpected error occurred: "https://registry.yarnpkg.com/async/-/async-1.5.2.tgz: EEXIST: file already exists, mkdir '/home/buildbot/.cache/yarn/v1/npm-async-1.5.2-ec6a61ae56480c0c3cb241c95618e20892f9672a'".
...
```

File: master/buildbot/newsfragments/unauthorized-login.bugfix [hunk 3]
```diff
@@ -0,0 +1 @@
+Fixed UnauthorizedLogin errors during buildbot restart. Fixes (:issue:`3462`).
```
Comment (tardyp) [inline master/buildbot/newsfragments/unauthorized-login.bugfix]: I think warn here that there is a new lock.

    Worker authentication is now delayed via a DeferredLock until Buildbot configuration is finished. This fixes UnauthorizedLogin errors during buildbot restart (:issue:`3462`)

Comment (rjarry) [reply master/buildbot/newsfragments/unauthorized-login.bugfix]: I updated the news fragment accordingly :+1:

**Comment by rjarry** [comment]
thanks :)

**Pull request merged**